### DISCUSSION TITLE
IPOPT interface, OPTION_INVALID exception

### DISCUSSION THREAD

**Discussion by jdeschut** [casadi/casadi#3938] [General]
Dear developers,

I am solving NLPs with Casadi 3.6.4 and occasionally, I get the following strange exception:

```
Exception of type: OPTION_INVALID in file "/work/build/external_projects/src/ipopt-external/src/Common/IpOptionsList.cpp" at line 558:
 Exception message: Option "mu_init": Double value expected, but non-numeric value "0.001" found.
 ```
 
- I set this option as `opts['ipopt.mu_init'] = 1e-3`
 - I get the same error for the Option `tol`.
 - The problem only occurs for some initializations / NLP parameters.
 - Sometimes, the NLP solver is called succesfully once, and then, in a second call, throws this exception.
 
Any idea what could be going on?

Thanks and kind regards,
Jochem

**Comment by jdeschut** [comment]
Update: in one instance I was using a `copy.deepcopy()` of some other options `dict` as an input to `nlpsol`. By working around copying the dict, the problem disappeared. Still strange that it works in some arbitrary cases and in some not.

**Comment by jgillis** [comment]
Any failing case you can share?

**Comment by jdeschut** [comment]
It's based on a large CasADi-based codebase. With pickling the NLP functions and numerical values around the solver call, I'm currently unable to reproduce this exception. In any case, what fixed it for me was changing:

```
opts = {}
opts['ipopt.mu_init'] = 1e-3
opts_2 = copy.deepcopy(opts)
opts_2['ipopt.mu_target'] = 1e-3
opts_2['ipopt.tol'] = 1e-4
solver = ca.nlpsol('solver', 'ipopt', nlp, opts_2)

(... use opts somewhere else ...)
```

to

```
opts_2 = {}
opts_2['ipopt.mu_init'] = 1e-3
opts_2['ipopt.mu_target'] = 1e-3
opts_2['ipopt.tol'] = 1e-4
opts_2['ipopt.max_iter'] = 200
solver = ca.nlpsol('solver', 'ipopt', nlp, opts_2)
```
